pkg/autodiscovery/listeners/snmp.go

```
@@ -0,0 +1,351 @@
+type SNMPService struct {
+	adIdentifier string
+	entityID     string
+	deviceIP     string
+	creationTime integration.CreationTime
+}
+
+// Make sure SNMPService implements the Service interface
+var _ Service = &SNMPService{}
+
+type snmpSubnet struct {
+	adIdentifier  string
+	config        util.SNMPConfig
+	defaultParams *gosnmp.GoSNMP
+	startingIP    net.IP
+	currentIP     net.IP
+	network       net.IPNet
+	cacheKey      string
+}
+
+// NewSNMPListener creates a SNMPListener
+func NewSNMPListener() (ServiceListener, error) {
+	var snmpConfig util.SNMPListenerConfig
+	err := config.Datadog.UnmarshalKey("snmp_listener", &snmpConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &SNMPListener{
+		services: map[string]Service{},
+		devices:  map[string][]string{},
+		stop:     make(chan bool),
+		config:   snmpConfig,
+	}, nil
+}
+
+// Listen periodically refreshes devices
+func (l *SNMPListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go l.checkDevices()
+}
+
+func (l *SNMPListener) loadCache(config util.SNMPConfig, adIdentifier string, cacheKey string) {
+	cacheValue, err := persistentcache.Read(cacheKey)
+	if err != nil {
+		log.Errorf("Couldn't read cache for %s: %s", cacheKey, err)
+		return
+	}
+	if cacheValue == "" {
+		return
+	}
+	var devices []net.IP
+	err = json.Unmarshal([]byte(cacheValue), &devices)
+	if err != nil {
+		log.Errorf("Couldn't unmarshal cache for %s: %s", cacheKey, err)
+		return
+	}
+	for _, deviceIP := range devices {
+		entityID := config.Digest(deviceIP.String())
+		l.createService(deviceIP.String(), adIdentifier, entityID)
+	}
+}
+
+func writeCache(cacheKey string, devices []string) {
+	cacheValue, err := json.Marshal(devices)
+	if err != nil {
+		log.Errorf("Couldn't marshal cache: %s", err)
+		return
+	}
+	err = persistentcache.Write(cacheKey, string(cacheValue))
+	if err != nil {
+		log.Errorf("Couldn't write cache: %s", err)
+	}
```

```suggestion
	if err = persistentcache.Write(cacheKey, string(cacheValue)); err != nil {
		log.Errorf("Couldn't write cache: %s", err)
	}
```
but I'd rather return the error itself
`return persistentcache.Write(cacheKey, string(cacheValue))`